espnet2/asr/transducer/beam_search_transducer.py

```diff
@@ -0,0 +1,679 @@
+
+        nbest_hyps = self.search_algorithm(h)
+
+        return nbest_hyps
+
+    def sort_nbest(self, hyps: List[Hypothesis]) -> List[Hypothesis]:
+        """Sort hypotheses by score or score given sequence length.
+
+        Args:
+            hyps: list of hypotheses
+
+        Return:
+            hyps: sorted list of hypotheses
+
+        """
+        if self.score_norm:
+            return sorted(hyps, key=lambda x: x.score / len(x.yseq), reverse=True)
+        else:
+            return sorted(hyps, key=lambda x: x.score, reverse=True)
+
+    def greedy_search(self, h: torch.Tensor) -> List[Hypothesis]:
+        """Greedy search implementation for transformer-transducer.
+
+        Args:
+            h: Encoded speech features (T_max, D_enc)
+
+        Returns:
+            hyp: 1-best decoding results
+
+        """
+        init_tensor = h.unsqueeze(0)
+        dec_state = self.decoder.init_state(init_tensor)
+
+        hyp = Hypothesis(score=0.0, yseq=[self.blank], dec_state=dec_state)
+
+        cache = {}
+
+        y, state, _ = self.decoder.score_transducer(hyp, cache, init_tensor)
+
+        for i, hi in enumerate(h):
+            ytu = torch.log_softmax(self.joint_network(hi, y[0]), dim=-1)
+            logp, pred = torch.max(ytu, dim=-1)
+
+            if pred != self.blank:
+                hyp.yseq.append(int(pred))
+                hyp.score += float(logp)
+
+                hyp.dec_state = state
+
+                y, state, _ = self.decoder.score_transducer(hyp, cache, init_tensor)
+
+        return [hyp]
+
+    def default_beam_search(self, h: torch.Tensor) -> List[Hypothesis]:
+        """Beam search implementation.
+
+        Args:
+            x: Encoded speech features (T_max, D_enc)
+
+        Returns:
+            nbest_hyps: N-best decoding results
+
+        """
+        beam = min(self.beam_size, self.vocab_size)
+        beam_k = min(beam, (self.vocab_size - 1))
+
+        init_tensor = h.unsqueeze(0)
+        blank_tensor = init_tensor.new_zeros(1, dtype=torch.long)
+
+        dec_state = self.decoder.init_state(init_tensor)
+
+        kept_hyps = [Hypothesis(score=0.0, yseq=[self.blank], dec_state=dec_state)]
+
+        cache = {}
+
+        for hi in h:
+            hyps = kept_hyps
+            kept_hyps = []
+
+            while True:
+                max_hyp = max(hyps, key=lambda x: x.score)
+                hyps.remove(max_hyp)
+
+                y, state, lm_tokens = self.decoder.score_transducer(
+                    max_hyp, cache, init_tensor
+                )
+
+                ytu = torch.log_softmax(self.joint_network(hi, y[0]), dim=-1)
+
+                top_k = ytu[1:].topk(beam_k, dim=-1)
+
+                ytu = (
+                    torch.cat((top_k[0], ytu[0:1])),
+                    torch.cat((top_k[1] + 1, blank_tensor)),
+                )
+
+                if self.lm:
+                    lm_state, lm_scores = self.lm.predict(max_hyp.lm_state, lm_tokens)
```

Comment:
`predict` doesn't exist.

Reply:
Sorry, I didn't update LM part yet!

Reply:
Maybe, can we use lm as it is? LM.score can be used?

Reply:
I'm not familiar with the LM part in espnet2 sorry, I'll take a look and update!

Reply:
We can use `score` and `batch_score` indeed, I just need to add a `zero_state` method to LM class.
Btw, I'm wondering two things about:

- The LM implementation in espnet2 is more suited for transducer as I don't need to create batch states manually nor extra manipulation to extract specific LM state. I'm wondering how efficient is it compared to implementation in espnet1?
- Is there a reason we didn't rely on list of batches (L, hid_units) in `RNNDecoder` instead of list of layers (B, hid_units)? I know it's direction against another but it would be easier/cleaner in transducer case

Reply:
> * Is there a reason we didn't rely on list of batches x (L, hid_units) in RNNDecoder instead of list of layers x (B, hid_units)? I know it's a direction against another but I think it would be easier/cleaner in transducer case

What part are you talking about?

Reply:
Sorry it was confusing: I was talking about how we deal with LM state during inference in espnet2 vs espnet1 and expected input/output (i.e.: predict in espnet1 vs score in espnet2) . I was wondering if I could follow the same direction for decoder states and also which one was the most efficient in terms of speed in my case.

But don't mind the second question, I'll do some testing and benchmarking!